### ISSUE TITLE
Optional arguments sometimes unable to infer

### ISSUE THREAD

**Issue by unional** [microsoft/TypeScript#53792]
# Bug Report

### 🔎 Search Terms

infer optional arguments parameters.

Here are some issues that might be related:

- #33852
- #45425
- #51377
- #51925
- #53614

But this case is even simpler. 

### 🕗 Version & Regression Information

4.8.4 - 5.0.4

It is working correctly in 4.7.4

- This changed between versions 4.7.4 and 4.8.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/GYVwdgxgLglg9mABAEwKbBmVAeACo1ADylTGQGdFwBrMOAdzAG0BdAGkQCVVyQAbKAD4AFMDAAuRMIB0sgIYAnAOblJuAJSIAvIK49+UdZKgBPAA6o4wRGMQBvRAtRQQCpLYC+AWABQv3wD0AVJwZrAIcnwA-JIOcpJgIAC2AEaoCgDciB6IAD5UZOiYqMiaOogAbnAwyL4QCORQNmAAjNooRVjCwqHhYJEx9ojxiImp6dllug4e6v4+QVJTQ8BwcMJGldXIWd4+9WCNzQBM7WgYXT1h8P3RscMJyWkKk9q6wg6r65ozk3M+QA)

### 💻 Code

```ts
function define<P extends unknown[], Result>(fn: (...args: P) => Result): typeof fn { return fn }


// (optional?: { a: number; } | undefined) => void
const fn1 = define((optional?: { a: number }) => { })

// () => { foo(): void; }
const fn2 = define((optional?: { a: number }) => ({ foo() { } }))
```

### 🙁 Actual behavior

```ts
// () => { foo(): void; }
const fn2 = define((optional?: { a: number }) => ({ foo() { } }))
```

### 🙂 Expected behavior

```ts
// (optional?: { a: number; } | undefined) => { foo(): void; }
const fn2 = define((optional?: { a: number }) => ({ foo() { } }))
```

**Comment by jakebailey** [comment]
Check nightly; the behavior should be fixed. (On my phone or I'd bisect it.)

**Comment by unional** [comment]
Yes, confirmed it is fixed in nightly 20130415

**Comment by fatcerberus** [comment]
> nightly 20130415

*2013?*
